### ISSUE TITLE
Compaction when there's null/duplicate pool keys

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#4666]
Repro is with Zed commit 1aec5dc.

While it's certainly not encouraged, right now the Zed tools allow pools to be populated such that the pool key values are absent and hence reflected as `null` in the min/max object metadata. For instance, this might occur if a user is ignorant of the defaults and/or the role of the pool key, they create a pool with default settings, and their data doesn't happen to have a field called `ts`. While it may not be justified to invest heavy effort to deliver best case performance from such a suboptimal config, the tools should probably do their best to minimize catastrophic side effects and/or detect the problem and inform the user so they don't think the tools are fundamentally broken.

The simplified lake manager delivered in #4655 is currently capable of behaving poorly when confronted with such a pool. Here's an example repro. The sample data at `s3://brim-sampledata/fdns/fdns-a.zng.gz` has records that look like this:

```
$ zq -Z 'head 1' fdns-a.zng.gz
{
    name: "0.202.98.34.bc.googleusercontent.com",
    timestamp: 2020-11-27T00:28:12Z,
    type: "a",
    value: 34.98.202.0
}
```

i.e., they lack a field called `ts`. If I load this into a pool with default settings, an initial run of `zed manage -once` compacts everything into larger objects (due to separate/known issue #4664), but on each subsequent run of `zed manage -once` it keeps rewriting every object in the pool yet again.

```
$ zed -version
Version: v1.8.1-34-g1aec5dc8

$ zed manage -once
{"level":"info","ts":1687367405.973526,"logger":"pool","msg":"updating pool","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60}
{"level":"info","ts":1687367474.1350038,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60,"runs_found":1,"objects_compacted":32}

$ zed manage -once
{"level":"info","ts":1687367495.021746,"logger":"pool","msg":"updating pool","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60}
{"level":"info","ts":1687367558.315904,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60,"runs_found":1,"objects_compacted":9}

$ zed manage -once
{"level":"info","ts":1687367559.729755,"logger":"pool","msg":"updating pool","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60}
{"level":"info","ts":1687367622.307743,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60,"runs_found":1,"objects_compacted":9}

$ zed manage -once
{"level":"info","ts":1687367657.77381,"logger":"pool","msg":"updating pool","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60}
{"level":"info","ts":1687367725.079944,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2RWU2tv6DXdsWJWoB6mZqxa9gKl","branch":"main","interval":60,"runs_found":1,"objects_compacted":9}
```

This does not happen with healthy data like the `s3://brim-sampledata/conn-sorted.zng.gz` used in #4664. In that case once the pool becomes stable, each subsequent run of `zed manage -once` is a no-op.

**Comment by philrz** [comment]
Upon further investigation, perhaps this phenomenon is not unique to absence of pool key values causing `null` min/max object metadata. It seems I can reproduce pretty much the same thing when there's lots of duplicate values of pool keys.

Using that same FDNS test data, here I've specified a proper field name as the pool key (`zed create -orderby timestamp:asc fdns`), but the nature of the data is such that there's lots of duplicates of that pool key value. After the first compaction runs, we can see how the min & max values of objects are often the same, indicating duplicate pool key values across adjacent objects.

```
$ zed query -Z 'from fdns@main:objects | cut min,max,size | sort min'
{
    min: 1677-09-21T00:12:43.145224192Z,
    max: 2020-11-27T00:33:08Z,
    size: 524345469
}
{
    min: 2020-11-27T00:33:08Z,
    max: 2020-11-27T00:52:21Z,
    size: 524300456
}
{
    min: 2020-11-27T00:52:21Z,
    max: 2020-11-27T01:11:29Z,
    size: 524378286
}
{
    min: 2020-11-27T01:11:29Z,
    max: 2020-11-27T01:31:36Z,
    size: 524297441
}
{
    min: 2020-11-27T01:31:36Z,
    max: 2020-11-27T01:53:55Z,
    size: 524344107
}
{
    min: 2020-11-27T01:53:55Z,
    max: 2020-11-27T02:13:33Z,
    size: 524383871
}
{
    min: 2020-11-27T02:13:33Z,
    max: 2020-11-27T02:29:11Z,
    size: 524383447
}
{
    min: 2020-11-27T02:29:11Z,
    max: 2020-11-27T02:51:09Z,
    size: 524288058
}
{
    min: 2020-11-27T02:51:09Z,
    max: 2020-11-27T03:12:06Z,
    size: 524309748
}
{
    min: 2020-11-27T03:12:06Z,
    max: 2020-11-27T22:22:08Z,
    size: 326585433
}
```

And, as before, each subsequent run of `zed manage -once` triggers a rewrite of every object via compaction.

```
$ zed manage -once
{"level":"info","ts":1687371554.6339772,"logger":"pool","msg":"updating pool","name":"fdns","id":"2RWdcdbKY6ZoYVrFf1OWvXKZSxv","branch":"main","interval":60}
{"level":"info","ts":1687371662.9053152,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2RWdcdbKY6ZoYVrFf1OWvXKZSxv","branch":"main","interval":60,"runs_found":1,"objects_compacted":10}

$ zed manage -once
{"level":"info","ts":1687371668.899498,"logger":"pool","msg":"updating pool","name":"fdns","id":"2RWdcdbKY6ZoYVrFf1OWvXKZSxv","branch":"main","interval":60}
{"level":"info","ts":1687371777.846467,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2RWdcdbKY6ZoYVrFf1OWvXKZSxv","branch":"main","interval":60,"runs_found":1,"objects_compacted":10}
```

**Comment by philrz** [comment]
It looks like this was fixed by the changes in #4661.

Repeating the repro steps from the last comment now at Zed commit b566f6d, we no longer see the matching min & max values after the first compaction that we did before, and subsequent compaction runs exit immediately, implying they saw no work to do, as would be expected since no more data was added.

```
$ zed -version
Version: v1.8.1-75-gb566f6d5

$ zed create -orderby timestamp:asc fdns
pool created: fdns 2T4y9UNlorsxfEgGBsextZLXubb

$ zed use fdns
Switched to branch "main" on pool "fdns"

$ zed load fdns-a.zng.gz(1/1) 3.00GB/3.00GB 0B/s 100.00%
2T50GqiuZvwRLpIOKApuZc6LcZa committed

$ zed manage
{"level":"info","ts":1690317545.176851,"logger":"pool","msg":"updating pool","name":"fdns","id":"2T4y9UNlorsxfEgGBsextZLXubb","branch":"main","interval":60}
{"level":"info","ts":1690317792.2101219,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2T4y9UNlorsxfEgGBsextZLXubb","branch":"main","interval":60,"runs_found":1,"objects_compacted":32}

$ zed query -Z 'from fdns@main:objects | cut min,max,size | sort min'
{
    min: 1677-09-21T00:12:43.145224192Z,
    max: 2020-11-27T00:33:08Z,
    size: 524643541
}
{
    min: 2020-11-27T00:33:09Z,
    max: 2020-11-27T00:52:22Z,
    size: 524709901
}
{
    min: 2020-11-27T00:52:23Z,
    max: 2020-11-27T01:11:31Z,
    size: 524604241
}
{
    min: 2020-11-27T01:11:32Z,
    max: 2020-11-27T01:31:39Z,
    size: 524534490
}
{
    min: 2020-11-27T01:31:40Z,
    max: 2020-11-27T01:53:58Z,
    size: 524782602
}
{
    min: 2020-11-27T01:53:59Z,
    max: 2020-11-27T02:13:35Z,
    size: 524367047
}
{
    min: 2020-11-27T02:13:36Z,
    max: 2020-11-27T02:29:13Z,
    size: 524386930
}
{
    min: 2020-11-27T02:29:14Z,
    max: 2020-11-27T02:51:13Z,
    size: 524339509
}
{
    min: 2020-11-27T02:51:14Z,
    max: 2020-11-27T03:12:10Z,
    size: 524672908
}
{
    min: 2020-11-27T03:12:11Z,
    max: 2020-11-27T22:22:08Z,
    size: 324520480
}

$ zed manage
{"level":"info","ts":1690318021.0469,"logger":"pool","msg":"updating pool","name":"fdns","id":"2T4y9UNlorsxfEgGBsextZLXubb","branch":"main","interval":60}
```

For good measure I also repeated the repro steps from the issue's intro text where I use the pool defaults and hence effectively have no pool key working for me. In this case compaction lands all the data in one large object, which seems ok given this is a degenerate case. Most importantly, once again we find that a re-run of `zed manage` is effectively a no-op.

```
$ zed create -use fdns
pool created: fdns 2T51iCzLrW9h5O7A705qQqFERk9
Switched to branch "main" on pool "fdns"

$ zed load fdns-a.zng.gz 
(1/1) 3.00GB/3.00GB 0B/s 100.00%
2T539i0LQbV6ZD8XypLh6nDEbzm committed

$ zed manage
{"level":"info","ts":1690320300.226828,"logger":"pool","msg":"updating pool","name":"fdns","id":"2T51iCzLrW9h5O7A705qQqFERk9","branch":"main","interval":60}
{"level":"info","ts":1690320396.324033,"logger":"pool","msg":"compaction completed","name":"fdns","id":"2T51iCzLrW9h5O7A705qQqFERk9","branch":"main","interval":60,"runs_found":1,"objects_compacted":32}

$ zed query -Z 'from fdns@main:objects | cut min,max,size | sort min'
{
    min: null,
    max: null,
    size: 4495845401
}

$ zed manage
{"level":"info","ts":1690320435.48992,"logger":"pool","msg":"updating pool","name":"fdns","id":"2T51iCzLrW9h5O7A705qQqFERk9","branch":"main","interval":60}
```

Thanks @mattnibs!
